### PR TITLE
Feature/handle product deletions

### PR DESCRIPTION
**Pull request by hannunosto**
Handle product deletions.

## Description
* Add field `is_deleted` for Nosto product index table 
* `Invalidate` indexer will set this to true for the products that cannot be fetched anymore from Magento
* Sync indexer will discontinue products in Nosto & remove deleted ones from Nosto product index in Magento   

## How Has This Been Tested?
Tested locally by deleting products and running the indexers.

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] I have assigned the correct milestone or created one if non-existent.
- [x] I have correctly labeled this pull request.
- [x] I have linked the corresponding issue in this description.
- [x] I have updated the corresponding Jira ticket.
- [x] I have requested a review from at least 2 reviewers
- [x] I have checked the base branch of this pull request
- [x] I have checked my code for any possible security vulnerabilities

### CODE REVIEW

File: Controller/Export/Base.php [hunk 1]
```diff
@@ -136,4 +136,12 @@ public function export(AbstractCollection $collection)
         }
         return $result;
     }
+
+    /**
+     * @return NostoHelperScope
+     */
+    public function getNostoHelperScope(): NostoHelperScope
```
Comment (supercid) [inline Controller/Export/Base.php]: Should we use type hints from now on?

File: Model/Indexer/Data.php [hunk 2]
```diff
@@ -120,15 +111,18 @@ public function execute($ids)
 
     /**
      * @param array $ids
-     * @return void
+     * @return IndexCollection
```
Comment (supercid) [inline Model/Indexer/Data.php]: I'd add a quick comment `Should be dirty and not deleted`

File: Model/Product/Index/IndexRepository.php [hunk 3]
```diff
@@ -178,52 +155,10 @@ public function getByProductIdAndStoreId(int $productId, int $storeId)
...
-        $searchResults->setTotalCount($collection->getSize());
-
-        return $searchResults;
-    }
-
     /**
      * @param SearchCriteriaInterface $searchCriteria
      *
-     * @return SearchResult
+     * @return IndexSearchResults
```
Comment (supercid) [inline Model/Product/Index/IndexRepository.php]: Double-check this. `RepositoryUtil::search` is maked to return `Magento\Framework\Api\Search\SearchResult` to me.
Perhaps: `@return SearchResult|IndexSearchResults`

File: Model/Service/Index.php [hunk 4]
```diff
@@ -315,4 +336,105 @@ public function markAsInSyncProductByIdAndStore($productId, $storeId)
...
+
+    /**
+     * Fetches deleted products from the product index, sends those to Nosto
+     * and deletes the deleted rows from database
+     *
+     * @param Store $store
+     * @return int
+     * @throws NostoException
+     */
+    public function purgeDeletedProducts(Store $store)
```
Comment (supercid) [inline Model/Service/Index.php]: Shouldn't this method be on the repository class?

File: Observer/Order/Save.php [hunk 5]
```diff
@@ -105,8 +113,9 @@ public function __construct(
         $this->logger = $logger;
         $this->moduleManager = $moduleManager;
         $this->nostoOrderBuilder = $orderBuilder;
+        $this->orderStatusBuilder = $orderStatusBuilder;
         $this->customerRepository = $customerRepository;
-        $this->indexer = $indexerRegistry->get(Indexer::INDEXER_ID);
+        $this->indexer = $indexerRegistry->get(InvalidateIndexer::INDEXER_ID);
         $this->nostoHelperScope = $nostoHelperScope;
```
Comment (supercid) [inline Observer/Order/Save.php]: `$this->nostoHelperScope` can be removed from this class

File: Observer/Order/Save.php [hunk 6]
```diff
@@ -207,19 +197,100 @@ private function handleInventoryLevelUpdate(NostoOrder $nostoOrder)
...
+     * @param AccountInterface $nostoAccount
+     * @param Store $store
+     */
+    private function sendNewOrder(Order $order, AccountInterface $nostoAccount, Store $store)
+    {
+        /** @var NostoCustomer $nostoCustomer */
+        $nostoCustomer = $this->customerRepository
+            ->getOneByQuoteId($order->getQuoteId());
+        $nostoCustomerId = null;
+        $nostoCutomerIdentifier = NostoOrderCreate::IDENTIFIER_BY_CID;
```
Comment (supercid) [inline Observer/Order/Save.php]: Typo `$nostoCutomerIdentifier` -> `$nostoCustomerIdentifier`

File: Controller/Export/Base.php (hunk 1 above)
Comment (hannunosto) [reply Controller/Export/Base.php]: We can, but only if the method doesn't return null

File: Model/Service/Index.php (hunk 4 above)
Comment (hannunosto) [reply Model/Service/Index.php]: Let's do the refactoring later. The whole service is already quite bloated and will be most likely splitted into a few service class.

Comment (supercid) [reply Model/Service/Index.php]: 👍

**Pull request merged**